### PR TITLE
Raven 3 migration

### PR DESCRIPTION
**Pull request by Gkolocsar**
I updated Raven to the latest stable version and updated the server for CommonJobsDemo (DB and client).

Can you please review @AlphaGit @andresmoschini @sofipacifico ?

Thanks

### CODE REVIEW

File: source/CommonJobs/CommonJobs.Application/ApplicantFlow/ApplicantEventType_Suggestions.cs [hunk 1]
```diff
@@ -58,7 +58,8 @@ select new
                                  Predefined = g.Any(x => x.Predefined)
                              };
 
-            TransformResults = (db, results) => results.Where(x => !string.IsNullOrWhiteSpace(x.Text)).Select(x => x);
+            //TODO TRANSFORM: Review this
+            //TransformResults = (db, results) => results.Where(x => !string.IsNullOrWhiteSpace(x.Text)).Select(x => x);
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Application/ApplicantFlow/ApplicantEventType_Suggestions.cs]: Is this working correctly? If so, we could probably remove the comment.

File: source/CommonJobs/CommonJobs.Application/ApplicantFlow/GetEventTypeSuggestionsQuery.cs [hunk 2]
```diff
@@ -23,42 +24,58 @@ public GetEventTypeSuggestionsQuery(string term, int maxSuggestions = 8)
...
+            //            .Where(x => !results.Contains(x))
+            //            .Take(MaxSuggestions - results.Count)
+            //            .ToList();
 
-                    results.AddRange(extraResults);
-                }
-            }
+            //        results.AddRange(extraResults);
+            //    }
+            //}
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Application/ApplicantFlow/GetEventTypeSuggestionsQuery.cs]: Remove if not used. Also, why changing this?

File: source/CommonJobs/CommonJobs.Application/EmployeeAbsences/EmployeeAbsences_Suggestions.cs [hunk 3]
```diff
@@ -50,7 +50,8 @@ select new
                                  Predefined = g.Any(x => x.Predefined)
                              };
 
-            TransformResults = (db, results) => results.Where(x => !string.IsNullOrWhiteSpace(x.Text)).Select(x => x);
+            //TODO TRANSFORM: Review this
+            //TransformResults = (db, results) => results.Where(x => !string.IsNullOrWhiteSpace(x.Text)).Select(x => x);
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Application/EmployeeAbsences/EmployeeAbsences_Suggestions.cs]: Same comment as above.

File: source/CommonJobs/CommonJobs.Application/EmployeeAbsences/GetSuggestions.cs [hunk 4]
```diff
@@ -23,42 +24,21 @@ public GetSuggestions(string term, int maxSuggestions = 8)
...
-                        .Where(x => !string.IsNullOrWhiteSpace(x))
-                        .Where(x => !results.Contains(x))
-                        .Take(MaxSuggestions - results.Count)
-                        .ToList();
+            var query = q.ToList();
 
-                    results.AddRange(extraResults);
-                }
-            }
+            var results = query.Select(x => x.Text).ToList();
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Application/EmployeeAbsences/GetSuggestions.cs]: Same comment as above.

File: source/CommonJobs/CommonJobs.Application/JobSearchSearching/GetJobSearchSuggestedApplicants.cs [hunk 5]
```diff
@@ -34,7 +35,8 @@ public override SuggestedApplicantsResult[] Execute()
                 .Statistics(out stats)
                 .Customize(x => x.WaitForNonStaleResultsAsOfLastWrite());
 
-            var result = query.AsProjection<SuggestedApplicantsResult>().ToArray();
+            //var result = query.AsProjection<SuggestedApplicantsResult>().ToArray();
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Application/JobSearchSearching/GetJobSearchSuggestedApplicants.cs]: Remove if not used.

File: source/CommonJobs/CommonJobs.Application/Suggest/GetSuggestions.cs [hunk 6]
```diff
@@ -25,43 +26,58 @@ public GetSuggestions(Expression<Func<Persons_Suggestions.Projection, object>> f
...
+            //            .Where(x => !results.Contains(x))
+            //            .Take(MaxSuggestions - results.Count)
+            //            .ToList();
 
-                    results.AddRange(extraResults);
-                }
-            }
+            //        results.AddRange(extraResults);
+            //    }
+            //}
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Application/Suggest/GetSuggestions.cs]: Same comment as above.

File: source/CommonJobs/CommonJobs.ContentExtraction.IFilterExtraction.Test/packages.config [hunk 7]
```diff
@@ -0,0 +1,4 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Newtonsoft.Json" version="7.0.1" targetFramework="net45" />
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.ContentExtraction.IFilterExtraction.Test/packages.config]: Is this needed here? Is it possible that this dependency has been added to more projects than actually need it?

File: source/CommonJobs/CommonJobs.Domain/Applicant.cs [hunk 8]
```diff
@@ -41,9 +41,9 @@ public void AddGeneralNote(string note, AttachmentReference attachment = null)
             });
         }
 
-        public override IEnumerable<SlotWithAttachment> AllAttachmentReferences
+        public override SlotWithAttachment[] AllAttachmentReferences
         {
-            get { return base.AllAttachmentReferences.Union(SlotWithAttachment.GenerateFromNotes(Notes)); }
+            get { return base.AllAttachmentReferences.Union(SlotWithAttachment.GenerateFromNotes(Notes)).ToArray(); }
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Domain/Applicant.cs]: `ToArray` will force the materialization of an enumeration, but `IEnumerable<>` will not, until iterated. Why the change?

File: source/CommonJobs/CommonJobs.Domain/Employee.cs [hunk 9]
```diff
@@ -159,14 +159,15 @@ public Decimal CurrentSalary
         [Display(Name = "Vencimiento del pasaporte europeo")]
         [DataType(DataType.Date)]
         public DateTime? EuropeanPassportExpiration { get; set; }
-        
-        public override IEnumerable<SlotWithAttachment> AllAttachmentReferences
+
+        public override SlotWithAttachment[] AllAttachmentReferences
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Domain/Employee.cs]: Same comment as above.

File: source/CommonJobs/CommonJobs.Domain/Person.cs [hunk 10]
```diff
@@ -121,9 +121,9 @@ public string Id
         //Only in detailed view
         public string Skills { get; set; }
 
-        public virtual IEnumerable<SlotWithAttachment> AllAttachmentReferences
+        public virtual SlotWithAttachment[] AllAttachmentReferences
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Domain/Person.cs]: Same comment as above. (I will mark them all so that later we can double check they're all resolved, one way or another.)

File: source/CommonJobs/CommonJobs.Domain/SlotWithAttachment.cs [hunk 11]
```diff
@@ -12,7 +12,7 @@ public class SlotWithAttachment
         public DateTime Date { get; set; }
         public AttachmentReference Attachment { get; set; }
 
-        public static IEnumerable<SlotWithAttachment> GenerateFromNotes(IEnumerable<NoteWithAttachment> notes)
+        public static SlotWithAttachment[] GenerateFromNotes(List<NoteWithAttachment> notes)
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Domain/SlotWithAttachment.cs]: Same comment as above.

File: source/CommonJobs/CommonJobs.Infrastructure.RavenDb/Command.cs [hunk 12]
```diff
@@ -3,7 +3,7 @@
 using System.Linq;
 using System.Web;
 using Raven.Client;
-using Newtonsoft.Json;
+using Raven.Imports.Newtonsoft.Json;
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Infrastructure.RavenDb/Command.cs]: Didn't you add the standalone Newtonsoft.Json?

File: source/CommonJobs/CommonJobs.Migrations/201212010811_CreateDefaultMenus.cs [hunk 13]
```diff
@@ -21,7 +21,8 @@ public override void Up()
         {
             var filePath = Path.Combine(Path.GetDirectoryName(Path.GetDirectoryName(new Uri(GetType().Assembly.CodeBase).LocalPath)), "Import201212010811.json");
             var serializer = new Newtonsoft.Json.JsonSerializer();
-            var placeByUserName = serializer.Deserialize <Dictionary<string, string>>(new StreamReader(filePath));
+            //TODO: VALIDATE THIS
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Migrations/201212010811_CreateDefaultMenus.cs]: What needs to be validated?

File: source/CommonJobs/CommonJobs.Migrations/201301161535_CleanLastModifiedBy.cs [hunk 14]
```diff
@@ -36,8 +36,12 @@ public override void Up()
 
             DocumentStore.DatabaseCommands.UpdateByIndex("Raven/DocumentsByEntityName",
                                                          new IndexQuery {  },
-                                                         patchs.ToArray(), 
-                                                         allowStale: true);
+                                                         patchs.ToArray(),
+                                                         //TODO: VALIDATE THIS
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Migrations/201301161535_CleanLastModifiedBy.cs]: What needs to be validated?

File: source/CommonJobs/CommonJobs.Mvc.UI/Controllers/ApplicantsController.cs [hunk 15]
```diff
@@ -40,13 +40,14 @@ public JsonNetResult List(ApplicantSearchParameters searchParameters)
...
             });
-        } 
+        }
 
         // GET: /Applicants/QuickSearchAutocomplete?terms=Mar
         public JsonResult QuickSearchAutocomplete(string term)
         {
             const int maxResults = 20;
-            var list = RavenSession.Advanced.DatabaseCommands
+            //TODO: Validate this!!!
```
Comment (AlphaGit) [inline source/CommonJobs/CommonJobs.Mvc.UI/Controllers/ApplicantsController.cs]: What needs to be validated?

File: source/CommonJobs/CommonJobs.Application/ApplicantFlow/GetEventTypeSuggestionsQuery.cs (hunk 2 above)
Comment (Gkolocsar) [inline source/CommonJobs/CommonJobs.Application/ApplicantFlow/GetEventTypeSuggestionsQuery.cs]: This is partial. The old query wasn't working with the new API so I had to rework it, but it's not good enough. This also happens with other "suggestions queries"

File: source/CommonJobs/CommonJobs.ContentExtraction.IFilterExtraction.Test/packages.config (hunk 7 above)
Comment (Gkolocsar) [inline source/CommonJobs/CommonJobs.ContentExtraction.IFilterExtraction.Test/packages.config]: It's possible, will review it.

File: source/CommonJobs/CommonJobs.Domain/Applicant.cs (hunk 8 above)
Comment (Gkolocsar) [inline source/CommonJobs/CommonJobs.Domain/Applicant.cs]: Compatibily issues with the new API. Couldn't make it work as it was.

File: source/CommonJobs/CommonJobs.Infrastructure.RavenDb/Command.cs (hunk 12 above)
Comment (Gkolocsar) [inline source/CommonJobs/CommonJobs.Infrastructure.RavenDb/Command.cs]: We should enforce the use of Raven's own newtonsoft json (raven has a copy of newtonsoft.json to avoid some known issues). So I will probably remove the plain Newtonsoft.json lib from where it's not needed.

File: source/CommonJobs/CommonJobs.Migrations/201212010811_CreateDefaultMenus.cs (hunk 13 above)
Comment (Gkolocsar) [inline source/CommonJobs/CommonJobs.Migrations/201212010811_CreateDefaultMenus.cs]: I didn't test migrations (not sure how). But I changed the StreamReader, so it was kind of a reminder.

File: source/CommonJobs/CommonJobs.Migrations/201301161535_CleanLastModifiedBy.cs (hunk 14 above)
Comment (Gkolocsar) [inline source/CommonJobs/CommonJobs.Migrations/201301161535_CleanLastModifiedBy.cs]: The new new BulkOperationOptions class passed as a parameter. Again, a reminder since I didn't test migrations.

File: source/CommonJobs/CommonJobs.Mvc.UI/Controllers/ApplicantsController.cs (hunk 15 above)
Comment (Gkolocsar) [inline source/CommonJobs/CommonJobs.Mvc.UI/Controllers/ApplicantsController.cs]: This comment can be removed. The new namespace is correct.
